src/sources/postgresql_metrics.rs

```diff
@@ -179,100 +183,131 @@ impl SourceConfig for PostgresqlMetricsConfig {
-            return Self::Exclude {
-                need_null,
-                databases,
-            };
+        let mut pg_stat_database_conflicts_sql =
+            "SELECT * FROM pg_stat_database_conflicts".to_owned();
+        if !match_sql.is_empty() {
+            pg_stat_database_conflicts_sql += " WHERE";
+            pg_stat_database_conflicts_sql += &match_sql;
         }
 
-        Self::None
+        Ok(Self {
+            pg_stat_database_sql,
+            pg_stat_database_conflicts_sql,
+            match_params,
+        })
     }
 
-    fn remove_empty(mut names: HashSet<String>) -> (bool, Vec<String>) {
-        let null = names.remove(&"".to_owned());
-        (null, names.into_iter().collect())
+    fn clean_databases(names: Vec<String>) -> (Vec<String>, bool) {
+        let mut set = names.into_iter().collect::<HashSet<_>>();
+        let null = set.remove(&"".to_owned());
+        (set.into_iter().collect(), null)
     }
 
-    async fn pg_stat_database(&self, client: &Client) -> Result<Vec<Row>, PgError> {
-        let mut conditions = "SELECT * FROM pg_stat_database".to_owned();
-        match self {
-            Self::Include {
-                need_null,
-                databases,
-            } => {
-                conditions += " WHERE datname = ANY($1)";
-                if *need_null {
-                    conditions += " OR datname IS NULL";
+    fn build_match_params(include: Vec<String>, exclude: Vec<String>) -> (String, Vec<String>) {
+        let mut query = String::new();
+        let mut params = vec![];
+
+        if !include.is_empty() {
+            query.push_str(" (");
+            for (i, name) in include.into_iter().enumerate() {
+                params.push(name);
+                if i > 0 {
+                    query.push_str(" OR");
                 }
-                client
-                    .query(conditions.as_str(), &[&databases.as_slice()])
-                    .await
+                query.push_str(&format!(" datname ~ ${}", params.len()));
             }
-            Self::Exclude {
-                need_null,
-                databases,
-            } => {
-                conditions += " WHERE datname != ANY($1)";
-                if *need_null {
-                    conditions += " AND datname IS NOT NULL";
-                } else {
-                    conditions += " OR datname IS NULL";
+            query.push(')');
+        }
+
+        if !exclude.is_empty() {
+            if !query.is_empty() {
+                query.push_str(" AND");
+            }
+
+            query.push_str(" (");
+            for (i, name) in exclude.into_iter().enumerate() {
+                params.push(name);
+                if i > 0 {
+                    query.push_str(" AND");
                 }
-                client
-                    .query(conditions.as_str(), &[&databases.as_slice()])
-                    .await
+                query.push_str(&format!(" datname !~ ${}", params.len()));
```

Comment:
Not a big deal, but I think you could make this more consistent by pulling the `NOT` up. So you end up with queries like:

`... WHERE (datname ~ 'foo' OR datname ~ 'bar') AND NOT (datname ~ 'foobar')`

Where the first grouping is the `include_databases` and the second is `exclude_databases`.